staging/src/k8s.io/component-helpers/resource/helpers.go

```diff
@@ -44,15 +45,80 @@ type PodResourcesOptions struct {
+	// is enabled.
+	PodLevelResourcesEnabled bool
+	// UseContainerLevelResources controls whether pod resource calculation should use
+	// container-level resources to calculate PodRequests even when
+	// PodLevelResources feature is enabled and pod-level resources are set.
+	UseContainerLevelResources bool
 }
 
-// PodRequests computes the pod requests per the PodResourcesOptions supplied. If PodResourcesOptions is nil, then
-// the requests are returned including pod overhead. The computation is part of the API and must be reviewed
-// as an API change.
+var supportedPodLevelResources = sets.NewString(string(v1.ResourceCPU), string(v1.ResourceMemory))
+
+// IsSupportedPodLevelResources checks if a given resource is supported by pod-level
+// resource management through the PodLevelResources feature. Returns true if
+// the resource is supported.
+func IsSupportedPodLevelResource(name v1.ResourceName) bool {
+	return supportedPodLevelResources.Has(string(name))
+}
+
+// IsPodLevelResourcesSet check if PodLevelResources feature is enabled and
+// pod-level resources are set.
+func IsPodLevelResourcesSet(pod *v1.Pod, podLevelResourcesEnabled bool) bool {
+	return podLevelResourcesEnabled && pod.Spec.Resources != nil
+}
+
+// IsPodLevelResourcesSet check if PodLevelResources feature is enabled and
+// pod-level requests are set.
+func IsPodLevelRequestsSet(pod *v1.Pod, podLevelResourcesenabled bool) bool {
+	return IsPodLevelResourcesSet(pod, podLevelResourcesenabled) && pod.Spec.Resources.Requests != nil
+}
+
+// PodRequests computes the total pod requests per the PodResourcesOptions supplied.
+// If PodResourcesOptions is nil, then the requests are returned including pod overhead.
+// If the PodLevelResources feature is enabled AND the pod-level resources are set,
+// those pod-level values are used in calculating Pod Requests.
+// The computation is part of the API and must be reviewed as an API change.
 func PodRequests(pod *v1.Pod, opts PodResourcesOptions) v1.ResourceList {
 	// attempt to reuse the maps if passed, or allocate otherwise
 	reqs := reuseOrClearResourceList(opts.Reuse)
+	effectiveContainersReqs := effectiveContainersRequests(pod, opts)
+	if !opts.UseContainerLevelResources && IsPodLevelRequestsSet(pod, opts.PodLevelResourcesEnabled) {
+		podRequests := v1.ResourceList{}
+		for resourceName, quantity := range pod.Spec.Resources.Requests {
+			if IsSupportedPodLevelResource(resourceName) {
+				podRequests[resourceName] = quantity
+			}
+		}
 
+		// Total pod resource requests are calculated using the effective container
+		// requests when resources are not specified at pod-level or when resources
+		// are not supported by PodLevelResource feature
+		for resourceName, quantity := range effectiveContainersReqs {
+			if _, exists := podRequests[resourceName]; !exists {
+				podRequests[resourceName] = quantity
+			}
+		}
+		addResourceList(reqs, podRequests)
+	} else {
+		addResourceList(reqs, effectiveContainersReqs)
+	}
```

Comment:
I think this should reuse reqs, and then overwrite the values with the pod requests as needed. Otherwise it sort of defeats the purpose of avoiding allocating for reqs right? This also simplifies this method a lot.

```suggestion
	reqs := effectiveContainersRequests(pod, opts) // <-- do the clear/reuse in here
	if !opts.UseContainerLevelResources && IsPodLevelRequestsSet(pod, opts.PodLevelResourcesEnabled) {
		for resourceName, quantity := range pod.Spec.Resources.Requests {
			if IsSupportedPodLevelResource(resourceName) {
				reqs[resourceName] = quantity
			}
		}
	}
```

Reply:
Thank you for catching this. In attempt to amend to the existing logic, my brain couldn't think of this solution! Extremely valuable!